src/cs50/sql.py

```diff
@@ -316,8 +319,14 @@ def shutdown_session(exception=None):
                 # Join tokens into statement, abbreviating binary data as <class 'bytes'>
                 _statement = "".join([str(bytes) if token.ttype == sqlparse.tokens.Other else str(token) for token in tokens])
 
+                # If COMMIT or ROLLBACK, turn on autocommit mode
+                if command in ["COMMIT", "ROLLBACK"] and "TO" not in statement:
```

Comment:
Goal is to avoid ROLLBACK TO (because starting to autocommit after a ROLLBACK TO would be unexpected behavior; that syntax does not denote the end of a transaction). As far as I know, if ROLLBACK is present, the only valid SQL syntax is `ROLLBACK [WORK] TO [SAVEPOINT] identifier` (per [here](https://www.w3resource.com/mysql/mysql-transaction.php#SDT), [here](https://www.postgresql.org/docs/current/sql-rollback-to.html), and [here](https://www.sqlite.org/lang_savepoint.html)), and the syntax is strict enough that you can't really nest statements into a ROLLBACK. Additionally `execute` already filters against having more than one statement at a time.

The main improvement here, if any, would be checking against tokens instead of against the statement string, but even then I think such a change wouldn't really be necessary. The only other part of this syntax, per the specifications above, that can have the letters `TO` is the savepoint name -- and if we're using a savepoint name, then the keyword `TO` is required anyway. So we kind of have an invariant here that in any single SQL statement of the `ROLLBACK` type, the keyword `TO` must indicate a savepoint (even if somehow you're getting the savepoint name with a convoluted nested statement or if the savepoint name contains TO) -- which is exactly what we're trying to know. So I'm not sure it's such a fragile check, especially since it's a very readable way to get the information we need.

Reply:
Hm, okay, doesn't seem a problematic assumption, given the grammars. But given that `statement` is a `str` here, would it not be a bit more robust to do `"TO" in tokens` to at least ensure a full string match?

Reply:
Sure, I think that definitely makes sense.